parsers/ocaml.c

```diff
@@ -1060,15 +1126,54 @@ static void typeDecl (vString * const ident, ocaToken what)
 
 	case OcaIDENTIFIER:
 		addTag (ident, K_TYPE);
-		pushStrongContext (ident, ContextType);
-		requestStrongPoping ();
-		waitedToken = Tok_EQ;
-		comeAfter = &typeSpecification;
-		toDoNext = &tillTokenOrFallback;
+		// true type declaration
+		if (whatNext == Tok_EQ)
+		{
+			pushStrongContext (ident, ContextType);
+			requestStrongPoping ();
+			toDoNext = &typeSpecification;
+		}
+		else
+		{	// we're in a sig
+			toDoNext = &globalScope;
+		}
 		break;
 
 	default:
-		globalScope (ident, what);
+		globalScope (ident, what, whatNext);
+	}
+}
+
+/** handle 'val' signatures in sigs and .mli files
+  * val ident : String.t -> Val.t
+  * Eventually, this will do cool things to annotate
+  * functions with their actual signatures. But for now,
+  * it's basically globalLet */
+static void val (vString * const ident, ocaToken what,
+	ocaToken UNUSED (whatNext))
+{
+	switch (what)
+	{
+	case Tok_PARL:
+	case OcaKEYWORD_rec:
+		break;
+
+	case Tok_Op:
+		/* we are defining a new operator, it's a
+		 * function definition */
+		addTag (ident, K_VAL);
+		toDoNext = &globalScope;
+		break;
+
+	case Tok_Val:	/* Can be a weiiird binding, or an '_' */
+	case OcaIDENTIFIER:
+		addTag (ident, K_VAL);
+		toDoNext = &globalScope;	// sig parser ?
+		break;
```

Comment:
The reason I didn't merge it is because I was thinking about the possibility of extending the parser to be able to handle the special signature syntax that interfaces use. I'd prefer to keep it separate so that extension is easier to consider, despite being currently repetitive.